core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -49,89 +51,169 @@ protected ManageSnapshots self() {
-    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+    ValidationException.check(ops.current().snapshot(snapshotId) != null,
         "Cannot roll back to unknown snapshot id: %s", snapshotId);
 
-    operation = SnapshotManagerOperation.ROLLBACK;
+    this.managerOperation = SnapshotManagerOperation.ROLLBACK;
     this.targetSnapshotId = snapshotId;
 
     return this;
   }
 
   @Override
   public ManageSnapshots rollbackToTime(long timestampMillis) {
-    operation = SnapshotManagerOperation.ROLLBACK;
     // find the latest snapshot by timestamp older than timestampMillis
-    Snapshot snapshot = SnapshotUtil.findLatestSnapshotOlderThan(table, timestampMillis);
+    Snapshot snapshot = findLatestAncestorOlderThan(ops.current(), timestampMillis);
     Preconditions.checkArgument(snapshot != null,
         "Cannot roll back, no valid snapshot older than: %s", timestampMillis);
+
+    this.managerOperation = SnapshotManagerOperation.ROLLBACK;
     this.targetSnapshotId = snapshot.snapshotId();
 
     return this;
   }
 
   @Override
   public ManageSnapshots rollbackTo(long snapshotId) {
-    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+    TableMetadata current = ops.current();
+    ValidationException.check(current.snapshot(snapshotId) != null,
         "Cannot roll back to unknown snapshot id: %s", snapshotId);
-
-    ValidationException.check(SnapshotUtil.isCurrentAncestor(table, snapshotId),
+    ValidationException.check(
+        !SnapshotUtil.ancestorIds(current.currentSnapshot(), current::snapshot).contains(snapshotId),
         "Cannot roll back to snapshot, not an ancestor of the current state: %s", snapshotId);
     return setCurrentSnapshot(snapshotId);
   }
 
+  private void validate(TableMetadata base) {
+    validateNonAncestor(base, targetSnapshotId);
+    WapUtil.validateWapPublish(base, targetSnapshotId);
+  }
+
+  @Override
+  public List<ManifestFile> apply(TableMetadata base) {
+    // this apply method is called by SnapshotProducer, which refreshes the current table state
+    // because the state may have changed in that refresh, the validations must be done here
+    validate(base);
+    return super.apply(base);
+  }
+
   @Override
   public Snapshot apply() {
+    TableMetadata base = ops.refresh();
+
     if (targetSnapshotId == null) {
       // if no target snapshot was configured then NOOP by returning current state
-      return table.currentSnapshot();
+      return base.currentSnapshot();
     }
-    switch (operation) {
+
+    switch (managerOperation) {
       case CHERRYPICK:
-        WapUtil.validateNonAncestor(this.table, this.targetSnapshotId);
-        WapUtil.validateWapPublish(this.table, this.targetSnapshotId);
-        return super.apply();
+        if (base.currentSnapshot().snapshotId() == base.snapshot(targetSnapshotId).parentId()) {
+          // the snapshot to cherrypick is already based on the current state: fast-forward
+          validate(base);
```

Comment:
how many times do you reckon `validate(base)` would be called per snapshot commit? I'm asking coz it's an expensive operations, validate is θ(2N) where N is # ancestors, which can grow monotonically.

Reply:
This PR only calls validate(base) once for each commit in most cases unless the picked snapshot is already based off current state. So this PR doesn't add any further complexity to what we already had earlier in the original changes. Although, i'm just pondering on how this can impact commit performance more generally. (this doesn't have to be addressed in this pr)

Reply:
I don't think we need to be concerned about performance here. Tables usually have under a few thousand snapshots for us, and we have very frequent commits and a 7 day retention window. Even if we process a few thousand snapshots a couple times, it's not going to be much of a delay because these are all in memory.

Also, we have to process the snapshots for each new version because the set of valid snapshots may have changed. Every time the commit fails, this will retry and validate the latest table state, which is unavoidable. In the best case, that will happen just once, but it could be up to the number of retries configured for the table.